eclair-core/src/main/scala/fr/acinq/eclair/payment/send/PaymentInitiator.scala

```diff
@@ -185,53 +185,53 @@ object PaymentInitiator {
+                                          fallbackFinalExpiryDelta: CltvExpiryDelta = Channel.MIN_CLTV_EXPIRY_DELTA,
                                           routeParams: Option[RouteParams] = None) {
     val recipientNodeId = paymentRequest.nodeId
     val paymentHash = paymentRequest.paymentHash
 
     // We add one block in order to not have our htlcs fail when a new block has just been found.
-    def finalExpiry(currentBlockHeight: Long) = finalExpiryDelta.toCltvExpiry(currentBlockHeight + 1)
+    def finalExpiry(currentBlockHeight: Long) = paymentRequest.minFinalCltvExpiryDelta.getOrElse(fallbackFinalExpiryDelta).toCltvExpiry(currentBlockHeight + 1)
   }
 
   /**
-   * @param recipientAmount  amount that should be received by the final recipient (usually from a Bolt 11 invoice).
-   * @param paymentHash      payment hash.
-   * @param recipientNodeId  id of the final recipient.
-   * @param maxAttempts      maximum number of retries.
-   * @param finalExpiryDelta expiry delta for the final recipient.
-   * @param paymentRequest   (optional) Bolt 11 invoice.
-   * @param externalId       (optional) externally-controlled identifier (to reconcile between application DB and eclair DB).
-   * @param assistedRoutes   (optional) routing hints (usually from a Bolt 11 invoice).
-   * @param routeParams      (optional) parameters to fine-tune the routing algorithm.
-   * @param userCustomTlvs   (optional) user-defined custom tlvs that will be added to the onion sent to the target node.
+   * @param recipientAmount          amount that should be received by the final recipient (usually from a Bolt 11 invoice).
+   * @param paymentHash              payment hash.
+   * @param recipientNodeId          id of the final recipient.
+   * @param maxAttempts              maximum number of retries.
+   * @param fallbackFinalExpiryDelta expiry delta for the final recipient when the [[paymentRequest]] doesn't specify it.
+   * @param paymentRequest           (optional) Bolt 11 invoice.
+   * @param externalId               (optional) externally-controlled identifier (to reconcile between application DB and eclair DB).
+   * @param assistedRoutes           (optional) routing hints (usually from a Bolt 11 invoice).
+   * @param routeParams              (optional) parameters to fine-tune the routing algorithm.
+   * @param userCustomTlvs           (optional) user-defined custom tlvs that will be added to the onion sent to the target node.
    */
   case class SendPaymentRequest(recipientAmount: MilliSatoshi,
                                 paymentHash: ByteVector32,
                                 recipientNodeId: PublicKey,
                                 maxAttempts: Int,
-                                finalExpiryDelta: CltvExpiryDelta = Channel.MIN_CLTV_EXPIRY_DELTA,
+                                fallbackFinalExpiryDelta: CltvExpiryDelta = Channel.MIN_CLTV_EXPIRY_DELTA,
```

Comment:
On second thought, I'm not too sure about this. The `Channel.MIN_CLTV_EXPIRY_DELTA` is spec-defined, that's why it is a constant (and not a configurable value). Why don't we change the definition of `PaymentRequest.minFinalCltvExpiryDelta` to :

```scala
lazy val minFinalCltvExpiryDelta: CltvExpiryDelta = tags.collectFirst {
    case cltvExpiry: PaymentRequest.MinFinalCltvExpiry => cltvExpiry.toCltvExpiryDelta
  } getOrElse(Channel.MIN_CLTV_EXPIRY_DELTA)
```

We could also keep the `PaymentRequest.minFinalCltvExpiryDelta` optional, and define a new :

```scala
lazy val minFinalCltvExpiryDelta: Option[CltvExpiryDelta] = ...
lazy val minFinalCltvExpiryDeltaWithFallback: CltvExpiryDelta = minFinalCltvExpiryDelta.getOrElse(Channel.MIN_CLTV_EXPIRY_DELTA)
```

But that seems overkill, since we can always iterate on the `tags` if we want to know whether the `minFinalCltvExpiryDelta` was explicitely defined in the payment request or not.

Reply:
> The Channel.MIN_CLTV_EXPIRY_DELTA is spec-defined, that's why it is a constant (and not a configurable value).

This is changing in the spec as we speak :)
We all agreed yesterday that hard-coding this in the spec was a bad idea from the start, and we want to remove it entirely.
The last commit of the spec PR now says that writers MUST specify a value in their invoice, and the default value of 18 is only there for backwards-compatibility for readers that receive invoices from legacy wallets/nodes that didn't specify this value.

The goal in the long(er) term is that this will allow us to change this value from an `Option[CltvExpiryDelta]` to a `CltvExpiryDelta` once enough nodes have upgraded.

Reply:
That's fine, let's redefine `Channel.MIN_CLTV_EXPIRY_DELTA` to `18` and take my first proposal, it is consistent with the decision to make writer `MUST` define the value?

Reply:
And we do need to keep this `fallbackFinalExpiryDelta` in these request classes to allow senders to tweak this.
Ideally when you receive an `expiryTooSoon` error you could guess that you need to provide a higher value here.

Reply:
Or at least the node operator can manually act on this error when the payment fails, whereas if we hide that inside `PaymentRequest.scala` we lose this ability.

Reply:
> And we do need to keep this fallbackFinalExpiryDelta in these request classes to allow senders to tweak this.

Why would senders need to tweak this?

> Ideally when you receive an expiryTooSoon error you could guess that you need to provide a higher value here.
> Or at least the node operator can manually act on this error when the payment fails, whereas if we hide that inside PaymentRequest.scala we lose this ability.

If the `minFinalExpiryTooSoon` is enforced in the payment request, then the only reason this would fail is a disagreement in the current block height, in that case the payment should be retried as-is? Anyway for the use-case you are describing, this isn't a fallback but an override.

Reply:
Senders may need to tweak this when paying invoices that don't specify this value.

I think older nodes will want to increase the values they accept (given the recent noise about flood and loot) which they can likely do in other implementations by just changing a configuration value, without specifying it in invoices (because the code isn't yet released in implementation to add this to invoices).

I agree that we should remove this fallback value at some point, but I think we should wait a bit before doing this because it removes flexibility...

Reply:
There are some unknowns that may play into this as well, like hodl invoices, so I'm a bit wary when making the API less flexible than before

Reply:
Looking deeper into this, I don't think we are handling `FinalIncorrectCltvExpiry` at all in the payment lifecycle?

We have four possible values for the `minFinalCltvExpiry`:
1) returned in a `FinalIncorrectCltvExpiry`
2) manual override (what we are discussing here)
3) in the payment request
4) default suggested value in the spec (9 blocks then, 18 blocks now)

Do we agree on the order of precedence? If we handle the `FinalIncorrectCltvExpiry` error correctly, I'm not sure we need the manual override.

Reply:
Thanks :), once lnd and c-lightning release support for setting it by default in invoices, I think we'll be able to safely remove those and simplify the code